File: Week_11/Practicum/Path with Maximum Probability.cpp
class Solution {
public:
    double maxProbability(int n, vector<vector<int>>& edges, vector<double>& succProb, int start_node, int end_node) {
        vector<vector<pair<int, double>>> graph(n, vector<pair<int, double>>());

        for(int j = 0; j < edges.size(); j++) {
            int u = edges[j][0];
            int v = edges[j][1];
            double prob = succProb[j];
            if(u != v){ 
                graph[u].push_back({v, prob}); 
                graph[v].push_back({u, prob});
            }
        }

        vector<double> maxProb(n, 0.0);
        maxProb[start_node] = 1.0;

        // Приоритетна опашка (max-heap) с вероятности като първи елемент
        priority_queue<pair<double, int>> pq;
        pq.push({1.0, start_node});

        while(!pq.empty()){
            auto [prob, u] = pq.top();
            pq.pop();

            // Ако достигнем до крайния връх, връщаме вероятността
            if(u == end_node)
                return prob;

            // Ако текущата вероятност е по-малка от вече записаната, пропускаме
            if(prob < maxProb[u])
                continue;

            // Разглеждаме всички съседни върхове на u
            for(auto &[v, p] : graph[u]){
                double newProb = prob * p;
                if(newProb > maxProb[v]){
                    maxProb[v] = newProb;
                    pq.push({newProb, v});
                }
            }
        }

        // Ако няма път до крайния връх, връщаме 0
        return 0.0;
    }
};

File: Week_11/Practicum/Path with Maximum Probability.py
from collections import defaultdict 


def dijkstra(graph, start, end):
    dist = {node: 0 for node in graph.keys()}
    dist[start] = 1

    pq = [(-1, start)]

    while len(pq):
        node_dist, node = heapq.heappop(pq)

        for neighb, prob in graph[node]:
            
            if prob * -node_dist > dist[neighb]:
                dist[neighb] = prob * -node_dist
                heapq.heappush(pq, (-dist[neighb], neighb))

    if end not in graph:
        return 0

    return dist[end]


class Solution:
    def maxProbability(self, n: int, edges: List[List[int]], succProb: List[float], start_node: int, end_node: int) -> float:
        graph = defaultdict(list)

        for edge, prob in zip(edges, succProb):
            graph[edge[0]].append((edge[1], prob))
            graph[edge[1]].append((edge[0], prob))

        return dijkstra(graph, start_node, end_node)